### PR TITLE
add get_feature_names_out to MinHashEncoder

### PR DESCRIPTION
**Pull request by LeoGrin**
We generate feature names as `feature_name_1_0`, ... `feature_name_1_n_components` etc. If feature names are missing, we use `x_0`, `x_1` etc
(+ a test to check that the MinHashEncoder is deterministic)

### CODE REVIEW

File: skrub/tests/test_minhash_encoder.py [hunk 1]
```diff
@@ -261,3 +261,52 @@ def test_check_fitted_minhash_encoder() -> None:
...
+
+def test_deterministic():
+    """Test that the encoder is deterministic"""
+    # TODO: add random state to encoder
+    encoder1 = MinHashEncoder(n_components=4)
+    encoder2 = MinHashEncoder(n_components=4)
+    X = np.array(["a", "b", "c", "d", "e", "f", "g", "h"])[:, None]
+    encoded1 = encoder1.fit_transform(X)
+    encoded2 = encoder2.fit_transform(X)
+    assert np.array_equal(encoded1, encoded2)
```
Comment (Vincent-Maladiere) [inline skrub/tests/test_minhash_encoder.py:274]: Here, let's directly use:
```python
from numpy.testing import assert_array_equal

...

assert_array_equal(encoded_1, encoded_2)
```
Using numpy for asserting is handy because it comes with many flavours, like `assert_array_almost_equal`.

File: skrub/tests/test_minhash_encoder.py [hunk 2]
```diff
@@ -261,3 +261,52 @@ def test_check_fitted_minhash_encoder() -> None:
...
+    encoder = MinHashEncoder(n_components=4)
+    X = pd.DataFrame(
+        {
+            "col1": ["a", "b", "c", "d", "e", "f", "g", "h"],
+            "col2": ["a", "b", "c", "d", "e", "f", "g", "h"],
+        }
+    )
+    encoder.fit(X)
+    # columns names should be col1_0 etc for each column and each component
+    assert all(
```
Comment (Vincent-Maladiere) [inline skrub/tests/test_minhash_encoder.py:289]: Ditto, also let's define the expected columns array separately to ease readability

File: skrub/tests/test_minhash_encoder.py [hunk 3]
```diff
@@ -261,3 +261,52 @@ def test_check_fitted_minhash_encoder() -> None:
...
+                "col2_1",
+                "col2_2",
+                "col2_3",
+            ]
+        )
+    )
+
+    # Test that it works with a list of strings
+    encoder = MinHashEncoder(n_components=4)
+    X = np.array(["a", "b", "c", "d", "e", "f", "g", "h"]).reshape(-1, 1)
```
Comment (Vincent-Maladiere) [inline skrub/tests/test_minhash_encoder.py:307]: Nitpick: let's stick to either `reshape(-1, 1)` or `[:, None]`, still to ease readability

File: skrub/tests/test_minhash_encoder.py [hunk 4]
```diff
@@ -261,3 +261,52 @@ def test_check_fitted_minhash_encoder() -> None:
...
+    encoder.fit(X)
+    assert all(
```
Comment (Vincent-Maladiere) [inline skrub/tests/test_minhash_encoder.py:309]: Ditto

File: skrub/_minhash_encoder.py [hunk 5]
```diff
@@ -337,3 +348,39 @@ def transform(self, X) -> np.ndarray:
...
 
         return X_out.astype(np.float64)  # The output is an int32 before conversion
+
+    def get_feature_names_out(self, input_features=None):
+        """Get output feature names for transformation.
+
+        The output feature names look like:
+        `["x0_0", "x0_1", ..., "x0_(n_components - 1)",
+        "x1_0", ..., "x1_(n_components - 1)", ...,
+        "x(n_features_out - 1)_(n_components - 1)"]`
```
Comment (Vincent-Maladiere) [inline skrub/_minhash_encoder.py:358]:
```suggestion
        ``["x0_0", "x0_1", ..., "x0_(n_components - 1)",
        "x1_0", ..., "x1_(n_components - 1)", ...,
        "x(n_features_out - 1)_(n_components - 1)"]``
```

File: skrub/_minhash_encoder.py [hunk 6]
```diff
@@ -337,3 +348,39 @@ def transform(self, X) -> np.ndarray:
...
+
+        Parameters
+        ----------
+        input_features : array-like of str or None, default=None
+            Input features.
+
+            - If `input_features` is `None`, then `feature_names_in_` is
```
Comment (Vincent-Maladiere) [inline skrub/_minhash_encoder.py:365]:
```suggestion
            - If ``input_features`` is ``None``, then ``feature_names_in_`` is
```

File: skrub/_minhash_encoder.py [hunk 7]
```diff
@@ -337,3 +348,39 @@ def transform(self, X) -> np.ndarray:
...
+              used as feature names in. If `feature_names_in_` is not defined,
```
Comment (Vincent-Maladiere) [inline skrub/_minhash_encoder.py:366]:
```suggestion
              used as feature names in. If ``feature_names_in_`` is not defined,
```

File: skrub/_minhash_encoder.py [hunk 8]
```diff
@@ -337,3 +348,39 @@ def transform(self, X) -> np.ndarray:
...
+              then the following input feature names are generated:
+              `["x0", "x1", ..., "x(n_features_in_ - 1)"]`.
```
Comment (Vincent-Maladiere) [inline skrub/_minhash_encoder.py:368]:
```suggestion
              ``["x0", "x1", ..., "x(n_features_in_ - 1)"]``.
```

File: skrub/_minhash_encoder.py [hunk 9]
```diff
@@ -337,3 +348,39 @@ def transform(self, X) -> np.ndarray:
...
+            - If `input_features` is an array-like, then `input_features` must
```
Comment (Vincent-Maladiere) [inline skrub/_minhash_encoder.py:369]:
```suggestion
            - If ``input_features`` is an array-like, then ``input_features`` must
```

File: skrub/_minhash_encoder.py [hunk 10]
```diff
@@ -337,3 +348,39 @@ def transform(self, X) -> np.ndarray:
...
+              match `feature_names_in_` if `feature_names_in_` is defined.
```
Comment (Vincent-Maladiere) [inline skrub/_minhash_encoder.py:370]:
```suggestion
              match ``feature_names_in_`` if ``feature_names_in_`` is defined.
```

**Review by Vincent-Maladiere — COMMENTED** [review]
Hey @LeoGrin, thanks for the PR! Hopefully, this will play nicely with TableVectorizer :)

Here are a couple of quick remarks:

File: skrub/_minhash_encoder.py [hunk 11]
```diff
@@ -244,6 +248,11 @@ def fit(self, X, y=None) -> "MinHashEncoder":
         :obj:`MinHashEncoder`
             The fitted :class:`MinHashEncoder` instance (self).
         """
+        self._check_n_features(X, reset=True)
+        self._check_feature_names(X, reset=True)
+        X = check_input(X)
+        self.n_features_in_ = X.shape[1]
```
Comment (Vincent-Maladiere) [inline skrub/_minhash_encoder.py:254]: This line can be removed since `self.n_features_in_` is already set in `self._check_n_features`. WDYT?

**Comment by LeoGrin** [comment]
Thanks for the comments @Vincent-Maladiere !

**Review by Vincent-Maladiere — APPROVED** [review]
Thanks @LeoGrin! LGTM after CI is green :)

File: skrub/tests/test_minhash_encoder.py [hunk 12]
```diff
@@ -261,3 +260,38 @@ def test_check_fitted_minhash_encoder() -> None:
...
+    encoder2 = MinHashEncoder(n_components=4)
+    X = np.array(["a", "b", "c", "d", "e", "f", "g", "h"])[:, None]
+    encoded1 = encoder1.fit_transform(X)
+    encoded2 = encoder2.fit_transform(X)
+    assert_array_equal(encoded1, encoded2)
+
+
+def test_get_feature_names_out():
+    """Test that get_feature_names_out returns the correct feature names"""
+    # encoder = MinHashEncoder(n_components=4)
```
Comment (jovan-stojanovic) [inline skrub/tests/test_minhash_encoder.py:278]:
```suggestion
```

File: skrub/tests/test_minhash_encoder.py [hunk 13]
```diff
@@ -261,3 +260,38 @@ def test_check_fitted_minhash_encoder() -> None:
     # Check that it works after fitting
     encoder.fit(X)
     encoder.transform(X)
+
+
+def test_deterministic():
+    """Test that the encoder is deterministic"""
+    # TODO: add random state to encoder
```
Comment (jovan-stojanovic) [inline skrub/tests/test_minhash_encoder.py:267]: Is there randomness? If it's deterministic, no need for random state.

**Review by jovan-stojanovic — CHANGES_REQUESTED** [review]
Thanks Leo! Two minor precision and this is good to be merged

File: skrub/tests/test_minhash_encoder.py (hunk 13 above)
Comment (LeoGrin) [reply skrub/tests/test_minhash_encoder.py:267]: Right now it's deterministic because the seed in the hash is fixed, but we could imagine allowing the user to vary the seed.

**Review by LilianBoulard — APPROVED** [review]
